pages/v4/Coroutines.md

```diff
@@ -0,0 +1,184 @@
+---
+title: "Coroutines"
+category: "basics"
+displayOrder: 10
+version: "v4"
+---
+
+First of all, what is a Coroutine?
+
+Taken from Wikipedia:
+
+_Coroutines are very similar to threads. However, coroutines are cooperatively multitasked, whereas threads are typically preemptively multitasked. This means that coroutines provide concurrency but not parallelism. The advantages of coroutines over threads are that they may be used in a hard-realtime context (switching between coroutines need not involve any system calls or any blocking calls whatsoever), there is no need for synchronisation primitives such as mutexes, semaphores, etc. in order to guard critical sections, and there is no need for support from the operating system._
+
+A Coroutine is a piece of code that executes "in parallel", over a certain amount of time, or generally asynchronously, with another; in our particular case, it is code that should run for an arbitrarily long amount of frames, without getting in the way of the usual flow of your code.
+
+This **does not mean** that the Coroutine is operating on a separate thread, and thus can be written ignoring the usual constraints of the Update-Draw loop of the game engine; each step of the Coroutine should be able to be performed in one Update call, so don't go around looking for prime numbers in a Coroutine, chances are your game would stutter.
```

Comment:
Sounds good, but could be more to-the-point: It isn't clear until the second paragraph that coroutines have nothing to do with threads. How about something like: "A Coroutine is a method that can be paused at a specific line, and continued with the next one the following frame", followed by some info on how they're not threads, can be useful for effects, game behaviors, cutscene-like scripting, etc.